sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -439,17 +445,17 @@ trait WritingSparkTask[W <: DataWriter[InternalRow]] extends Logging with Serial
     var count = 0L
     // write the data and commit this writer.
     Utils.tryWithSafeFinallyAndFailureCallbacks(block = {
-      while (iter.hasNext) {
+      val iterWithMetrics = iter.map { row =>
```

Comment:
can we new a java iterator to make the code clearer? The `count` is now in the closure, which is a bit tricky.